### ISSUE TITLE
About propagation

### ISSUE THREAD

**Issue by wangyu1213888** [bo-miao/HTR#2]
Hi, I would like to ask if using this mode of propagation to propagate masks at the beginning of training will lead to the propagation of masks that are basically foreground.

**Comment by bo-miao** [comment]
Hi,

Our method can use either a ground truth mask or a language-conditioned mask to learn temporal matching. I did not check the results in initial iteractions, but the model converges well. If you use a language-conditioned mask as a prior, please use .detach() to control backpropagation. Hope the source code below helps.

            # MASK PROPAGATION MODEL
            use_pred_mask = False  # b 1 h w
            if use_pred_mask:
                x = F.interpolate(out['pred_masks'][:, :1], scale_factor=4, mode='bilinear', align_corners=False)
                mem_mask = torch.sigmoid(x).detach()
            else:
                mem_mask, valid = nested_tensor_from_tensor_list([t["masks"] for t in targets], size_divisibility=32, split=False).decompose()
                mem_mask = mem_mask.to(samples.tensors)[:, :1]

            # features list [bt c h w], mask [b 1 h w], image [b t c h w]
            self.voshead.clean_memory()
            results_logits, results_prob = self.voshead([x.tensors for x in features[:-1]], mem_mask,
                                                        rearrange(samples.tensors, '(b t) c h w -> b t c h w', t=t), mem_gap=1)
            out['vos_logits'] = torch.stack(results_logits, dim=1)  # b t o h w
            out['vos_prob'] = torch.stack(results_prob, dim=1)  # b t o-1 h w
            out['run_ratio'] = run_ratio
            self.voshead.clean_memory()

**Comment by wangyu1213888** [comment]
Hi, I just wanted to ask you again, did you propagate the mask during the training phase? I also want to know where outputs['run_ratio'], "main_matcher_index"] and outputs["aux_matcher_index"] in the loss function came from. Or can the complete training process of this code be given? Thank you very much

**Comment by bo-miao** [comment]
Hi,

To train temporal propagation module, we need to propagate masks during training. 

The 'run_ratio' indicates the current training stage (e.g., if 10% of the training is completed, 'run_ratio' is 0.1) and is used for BootstrappedCE loss. For more details, I recommend reading papers and codes such as STCN in semi-supervised VOS.

The 'main_matcher_index' and 'aux_matcher_index' represent the best match indices obtained from Hungarian matching. Temporal propagation is performed only for the best query results to save memory and time.

Currently I do not plan to organize the code further because I am busy with several projects.

**Comment by wangyu1213888** [comment]
Thank you very much for your answer
